### PR TITLE
Reland async dispatch with workaround for hang.

### PR DESCRIPTION
**Pull request by jbaumanTT**
### Ticket
#15018 

### Problem description
Async dispatch allows us to dispatch program binaries before the previous program completed, but we were hitting some hangs on LLAMA and tt-train due to path reservation for mcast hanging.

### What's changed
When async dispatch was previously enabled, we hit a hang in LLAMA that seems
to occur when doing the path reserve for an mcast and a previous mcast is still
active. To work around this, force write barriers before mcasts that aren't
linked to earlier mcasts. The dispatcher is the only core that sends mcasts
over VC 5, and even with multiple CQs only one can be mcasting programs at a
time.

### Checklist
- [x] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [x] Model regression CI testing passes (if applicable)
- [x] Device performance regression CI testing passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) tests passes
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt_metal/impl/dispatch/command_queue.cpp [hunk 1]
```diff
@@ -1029,6 +1029,12 @@ void EnqueueProgramCommand::assemble_device_commands(
                         read_length = max_paged_length_per_sub_cmd;
                         write_length = read_length;
                     }
+                    if (!kernel_bins_dispatch_subcmds.back().empty()) {
+                        auto& back = kernel_bins_dispatch_subcmds.back().back();
+                        if (back.noc_xy_addr != noc_encoding) {
+                            back.flags = CQ_DISPATCH_CMD_PACKED_WRITE_LARGE_FLAG_UNLINK;
+                        }
```
Comment (tt-asaigal) [inline tt_metal/impl/dispatch/command_queue.cpp:1036]: To confirm, this is basically unlinking the mcasts for grid 0, when we switch to grid 1. How is this different from the previous case where we unlinked at the bottom of the `kernel_idx` loop?

File: tt_metal/impl/dispatch/command_queue.cpp [hunk 2]
```diff
@@ -1050,9 +1056,11 @@ void EnqueueProgramCommand::assemble_device_commands(
                 }
             }
         }
-        // Unlink the last subcmd of the current core range
-        if (!write_linear) {
-            kernel_bins_dispatch_subcmds.back().back().flags |= CQ_DISPATCH_CMD_PACKED_WRITE_LARGE_FLAG_UNLINK;
+    }
+    // Unlink the last subcmd of every dispatch, to ensure we don't hold the path reservation for an incredible long time.
```
Comment (tt-asaigal) [inline tt_metal/impl/dispatch/command_queue.cpp:1060]: Not unlinking here would have hung all subsequent path reservations issued by any other mcast in the dispatch. Should we mention this in the comment?

**Review by tt-asaigal — APPROVED** [review]

File: tt_metal/impl/dispatch/command_queue.cpp (hunk 1 above)
Comment (tt-asaigal) [reply tt_metal/impl/dispatch/command_queue.cpp:1036]: Actually, nvm I read your commit message. This makes sense.

**Review by abhullar-tt — APPROVED** [review]

File: tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/pgm_dispatch_golden.log [hunk 3]
```diff
@@ -1,7 +1,7 @@
...
-3.26, 3.28, 4.58, 5.21, 6.08, 6.09, 6.46, 7.19, 6.21, 6.15, 3.80, 7.63, 8.22, 14.02, 4.63, 4.64, 15.70, 29.68, 152.22, 177.60,
-3.56, 3.63, 5.47, 6.42, 7.71, 7.70,
+2.70, 2.70, 2.92, 3.02, 3.15, 3.18, 3.50, 3.53, 3.46, 5.90, 4.07, 4.45, 5.69, 12.25, 4.05, 4.15, 11.55, 12.29, 74.37, 85.31, 4.13, 14.16, 109.50,
+2.71, 2.71, 2.93, 3.05, 3.23, 3.23, 3.55, 3.56, 3.49, 5.91, 4.08, 4.46, 5.75, 12.28, 4.05, 4.16, 11.60, 12.38, 74.64, 85.66, 4.13, 14.23, 109.89,
+2.71, 2.71, 3.02, 3.23, 3.39, 3.39, 3.72, 3.81, 3.67, 5.93, 4.10, 4.51, 6.04, 12.48, 4.06, 4.18, 11.78, 12.55, 75.12, 86.29, 4.13, 14.46, 110.95,
+2.78, 2.78, 3.23, 3.40, 3.61, 3.63, 3.96, 3.94, 3.88, 6.00, 4.13, 4.64, 6.18, 12.73, 4.11, 4.21, 12.03, 13.06, 76.39, 87.99, 4.13, 15.40, 114.50,
+3.00, 3.00, 3.66, 3.98, 4.32, 4.35, 4.66, 4.69, 4.64, 6.22, 4.21, 5.05, 6.78, 13.41, 4.32, 4.35, 12.73, 17.65, 82.46, 95.97, 4.13, 19.60, 122.75,
+3.29, 3.29, 4.51, 5.12, 5.79, 5.79, 6.27, 7.16, 6.02, 6.49, 4.34, 7.78, 8.28, 14.76, 4.64, 4.59, 16.45, 29.33, 151.60, 176.73,
+3.57, 3.57, 5.40, 6.27, 7.17, 7.21,
```
[the remark this answers is not in the document]
Comment (jbaumanTT) [reply tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/pgm_dispatch_golden.log:7]: Perf is about 2% faster without barriers on average. Around 14 tests are better, the best getting around 30% faster. 

Some of the win here is linked change; comparing this change to one where I just enable async (no barriers, no linking improvement), I see this change is 0.7% slower than the plain async change. About 10 tests get more than 5% worse, and worst test case is around 19% worse. So the linking change saves about 1.3% average and 10% worst-case.

File: tt_metal/impl/dispatch/command_queue.cpp (hunk 2 above)
Comment (jbaumanTT) [reply tt_metal/impl/dispatch/command_queue.cpp]: Done.

File: tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/pgm_dispatch_golden.log (hunk 3 above)
[the remark this answers is not in the document]
Comment (tt-asaigal) [reply tests/tt_metal/tt_metal/perf_microbenchmark/dispatch/pgm_dispatch_golden.log:7]: Thanks John!

**Pull request merged**